modules/dnn/src/graph_simplifier.cpp

```diff
@@ -112,6 +112,23 @@ bool Subgraph::match(const Ptr<ImportGraphWrapper>& net, int nodeId,
                 continue;
             nodeId = getInputNodeId(net, node, j);
             const Ptr<ImportNodeWrapper> inpNode = net->getNode(nodeId);
+            if ((inpNode->getType() == "Gather") && (node->getType() == "Cast"))
```

Comment:
Why we need to change generic part of graphs pattern match?

Reply:
The pattern match is not correct when there are several Cast nodes that use Gather node as input. Right now, it merges Gather with one of the Cast nodes, and others no longer have a correct input node, so it causes an error.

Reply:
The thing is that this is a common graphs fusion part but "Gather" and "Cast" are too specific.

Reply:
This logic doesn't work specifically for the combination of "Gather" and "Cast" nodes. If you think that there is a better way for avoiding this error, please, share it, I will update the fix.

Reply:
ONNX related transformations should be located at https://github.com/opencv/opencv/blob/master/modules/dnn/src/onnx/onnx_graph_simplifier.cpp

Reply:
@sl-sergei `graph_simplifier.cpp` is a generic template. **Specialization** for specific cases should go to another place.